File: 2023-04(백준, 프로그래머스)/거리두기 확인하기.py
def solution(places):
    
    def check(arr,x,y):
        xy = [[x,y,0]]
        while xy:
            a,b,c = xy.pop()
            
            if c == 2:continue
            
            for i,j in [a+1,b],[a-1,b],[a,b-1],[a,b+1]:
                if 0 <= i < 5 and 0 <= j < 5 and arr[i][j] != 'X':
                    if arr[i][j] == 'P':
                        return False
                    xy.append([i,j,c+1])
        return True
    
    result = []
    
    for i in places:
        place = []
        p = []
        for x in range(5):
            temp = []
            for y in range(5):
                if i[x][y] == 'P':
                    p.append([x,y])
                temp.append(i[x][y])
            place.append(temp)
            
        for a,b in p:
            place[a][b] = 'X'
            if not check(place,a,b):
                result.append(0)
                break
            place[a][b] = 'P'
        else:
            result.append(1)
            
    return result

File: JS 코테 연습/거리두기 확인하기.js
function solution(places) {
    
    function check (board, q) {
        let temp = true
        
        while (q.length !== 0){
            let [x, y, cnt, visit] = q.shift()
            
            if (cnt === 3) continue
            
            if (cnt > 0 && board[x][y] === 'P'){
                temp = false
                break
            }
            
            for (let [i,j] of [[x+1,y], [x-1,y], [x,y+1], [x,y-1]]){
                let new_p = `${i}-${j}`
                if (0 <= i && i < 5 && 0 <= j && j < 5 && board[i][j] !== 'X' && !visit.includes(new_p)){
                    // 리스트에 추가시 ...를 사용하자
                    q.push([i,j,cnt+1, [new_p, ...visit]])
                }
            }
        }
        return temp
    }
    
    
    let result = []
    for (let place of places){
        let board = [...Array(5).keys()].map(_ => [...Array(5).keys()].map(l => 0))
        let position = []
        for (let i =0 ; i < 5; i++){
            for (let j=0; j < 5; j++){
                board[i][j] = place[i][j]
                if (place[i][j] === 'P'){
                    position.push([i, j, 0, [`${i}-${j}`]])
                }
            }
        }
        result.push(check(board, position) ? 1 : 0)
    }
    
    return result
}
solution(	[["POOOP", "OXXOX", "OPXPX", "OOXOX", "POXXP"], ["POOPX", "OXPXP", "PXXXO", "OXXXO", "OOOPP"], ["PXOPX", "OXOXP", "OXPOX", "OXXOP", "PXPOX"], ["OOOXX", "XOOOX", "OOOXX", "OXOOX", "OOOOO"], ["PXPXP", "XPXPX", "PXPXP", "XPXPX", "PXPXP"]])